### DISCUSSION TITLE
[RFC] Schema Load lifecycles - Compound Indexes and other schema modifications (#5972)

### DISCUSSION THREAD

**Discussion by diogotcorreia** [strapi/strapi#6681] [Legacy Topics]
Related issue: #5972 
Related PR: #6515 

---

It would be great if we could manually execute code against the schema object for either Mongoose or Bookshelf. That way, better customization would be possible.

This would have the added benefit of being able to add compound indexes, for example, which is not possible at the moment while using Mongoose.

---

### Drawbacks

This change would go against Strapi's goal of being database agnostic. However, I think there is a need for further customizing the integration without having to overwrite the dependencies using something like `patch-package`.

#### Idea 1

I'd say we could have another lifecycle in the `models.js` file, which would look like
```js
module.exports = {
  lifecycles: {
    onSchemaLoad(type, ...args) {
      // ...
    },
    beforeCreate(...args) { },
	// ...
};
```
where `type` would be a string indicating the type of database (`mongoose`, `postgresql`, `sqlite`, `mysql` or `mariadb`) - to be further discussed.

With this setup, models could have a lifecycle like this, which would allow for changes based on the current database.

```js
module.exports = {
  lifecycles: {
    onSchemaLoad(type, ...args) {
      if (type === "mongoose") {
		// do something
	  } else if (type === "sqlite") {
        // do something else
      }
    }
};
```

#### Idea 2

Another idea would be to have a function per database type:

```js
module.exports = {
  onLoad: {
    mongoose(...args) {
      // do something
    },
    sqlite(...args) {
      // do something else
    },
  },
  lifecycles: {
    // ...
  },
};
```

In both ideas, `...args` needs to be further discussed. However, for `mongoose` it should be the [`Schema`](https://mongoosejs.com/docs/api/schema.html) object. No word on `bookshelf` yet.

---

Comments and suggestions are appreciated.

**Comment by alexandrebodin** [comment]
Hi @diogotcorreia Sorry to be picky but I meant open an RFC on our RFC repo :D https://github.com/strapi/rfcs.

**Reply by diogotcorreia** [reply]
I thought that was dead since the last commit was 11 months ago. Will open something there.
